tests/Sales/OrderBuilderTest.php

```diff
@@ -0,0 +1,178 @@
+        self::assertNotEmpty($this->orderFixture->getOrderItemQtys());
+    }
+
+    /**
+     * Create an order for an internally generated customer.
+     *
+     * Control the product included with the order, use random item quantities.
+     *
+     * @test
+     * @throws \Exception
+     */
+    public function createOrderWithProduct()
+    {
+        $this->orderFixture = new OrderFixture(
+            OrderBuilder::anOrder()->withProducts(ProductBuilder::aSimpleProduct())->build()
+        );
+
+        self::assertInstanceOf(OrderInterface::class, $this->orderRepository->get($this->orderFixture->getId()));
+        self::assertCount(1, $this->orderFixture->getOrderItemQtys());
+    }
+
+    /**
+     * Create an order for an internally generated customer with multiple products.
+     *
+     * Control the products included with the order, use random item quantities.
+     *
+     * @test
+     * @throws \Exception
+     */
+    public function createOrderWithProducts()
+    {
+        $this->orderFixture = new OrderFixture(
+            OrderBuilder::anOrder()->withProducts(
+                ProductBuilder::aSimpleProduct()->withSku('foo'),
+                ProductBuilder::aSimpleProduct()->withSku('bar')
+            )->build()
+        );
+
+        self::assertInstanceOf(OrderInterface::class, $this->orderRepository->get($this->orderFixture->getId()));
+        self::assertCount(2, $this->orderFixture->getOrderItemQtys());
+    }
+
+    /**
+     * Create an order for a given customer with internally generated product(s).
+     *
+     * Control the customer placing the order.
+     *
+     * @test
+     * @throws \Exception
+     */
+    public function createOrderWithCustomer()
+    {
+        $customerEmail = 'test@example.com';
+        $customerBuilder = CustomerBuilder::aCustomer()
+            ->withEmail($customerEmail)
+            ->withAddresses(AddressBuilder::anAddress()->asDefaultBilling()->asDefaultShipping());
+
+        $this->orderFixture = new OrderFixture(
+            OrderBuilder::anOrder()->withCustomer($customerBuilder)->build()
+        );
+
+        self::assertInstanceOf(OrderInterface::class, $this->orderRepository->get($this->orderFixture->getId()));
+        self::assertSame($customerEmail, $this->orderFixture->getCustomerEmail());
+        self::assertNotEmpty($this->orderFixture->getOrderItemQtys());
+    }
+
+    /**
+     * Create an order for a given cart.
+     *
+     * Complex to set up, most flexible:
+     * - define products
+     * - define customer
+     * - set item quantities
+     * - set payment and shipping method
+     *
+     * @test
+     * @throws \Exception
+     */
+    public function createOrderWithCart()
+    {
+        $cartItems = ['foo' => 2, 'bar' => 3];
+        $customerEmail = 'test@example.com';
+        $paymentMethod = 'checkmo';
+        $shippingMethod = 'flatrate_flatrate';
+
+        $productBuilders = [];
```

Comment:
Aren't the product builders ignored if `withCart()` is used? Or is this what you test here by passing both? For clarity I would suggest to split that into a separate test then, e.g. `createOrderWithCartIgnoresProductBuilders()`

Reply:
Possible setups are:

* Omit product builders, omit cart → Generated products will be added to the catalog, all of them added to cart
* Specify product builders, omit cart → Given products will be added to the catalog, all of them added to cart
* Specify product builders, specify cart → Given products will be added to the catalog, given items added to cart

`withCart()` cannot be used without product builders. Cart items can be generated from product builders but catalog products cannot be built from the cart builder. Does that maybe clarify some of your remarks?

Reply:
Got it, so it's a shortcut to generate the products even if they are not added to the order. Any idea how that can be made more intuitive? This could be confusing to use.

Reply:
how about throwing an exception if cart builder was added but no product builder(s)?

Reply:
I'd say creating products by other means is still a valid use case